x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -240,7 +241,7 @@ private void getUserClaims(@Nullable AccessToken accessToken, JWT idToken, Nonce
                 }
                 claimsListener.onResponse(enrichedVerifiedIdTokenClaims);
             }
-        } catch (BadJOSEException e) {
```

Comment:
@jkakavas Can you please explain why `BadJOSEException` is not retry-able anymore?
If I read the code correctly, https://bitbucket.org/connect2id/nimbus-jose-jwt/src/8bc51863af3c36ec3631878ac69c2ab59936cfae/src/main/java/com/nimbusds/jose/proc/DefaultJOSEProcessor.java#lines-234 , a `BadJOSEException` is triggered when there is no candidate key to do the signature verification, so it must be the case of a key rotation?

Reply:
@albertzaharovits you are right about the above. The problem - which I tried to address - is that `BadJWTException` is also a `BadJOSEException` and that a `BadJWTException` can be thrown for many reasons ( Expired JWTs, wrong nonce, wrong issuer etc. etc. ) that have nothing to do with key rotation or reloadability of keys, so it didn't make sense to try and reload keys for any error that could happen in the context of the OIDC realm configuration or JWT parsing. 

I was under the impression that the error of not finding a suitable key would throw a `BadJWSException` and not a `BadJOSEException` but you are right. It feels like the only way to try and reload only on errors that can be caused by key rotation, we'd have to both match the Exception class _and_ the Exception message. This is what I was doing to start with here and [we agreed it was not an elegant solution](https://github.com/elastic/elasticsearch/pull/37787/files#r253671855)

Another solution would be to add the possibility to refresh a remote JWKSet at will on configurable intervals. This way, if you're using an OP that has a frequent key rotation schedule you can adjust the polling frequency accordingly. This fails to address the situation where an OP rotates their keys because of a compromise though - which is what the original implementation would try to cater for with the simplistic retry on error approach.

Reply:
@jkakavas 
> BadJWSException is also a BadJOSEException

I see, that I haven't noticed.

I think we might get a `BadJWSException` as well as a `BadJOSEException` when the OP rotated the keys, depending on if the OP uses the `kid` header parameter https://tools.ietf.org/html/rfc7515#section-4.1.4 . If it does I think the library would throw a `BadJOSEException`; if the OP simply updates the key we will get a `BadJWSException`.
Therefore, I think we should reload for the `BadJOSEException` even though `BadJOSEException` can also be caused by a user typo in the realm configuration. If we are concerned about doubling error messages for user misconfiguration, then we can only proceed with the claim validation retrial if the newly retrieved keys are different. I would not be concerned that a realm misconfiguration incurs an additional call to the OP.
What do you think?

Reply:
> @jkakavas
>    BadJWSException is also a BadJOSEException
> I see, that I haven't noticed.

It's BadJWTException being a BadJOSEException that is problematic, tbc. 

> I would not be concerned that a realm misconfiguration incurs an additional call to the OP.

Agreed, this is better than not reloading keys in case of a valid key rotation - this was an unintended consequence of the changes in this PR. I failed to notice that the library throws a `BadJOSEException` in some cases too. 

I'll raise a PR to change this

Reply:
Addressed in https://github.com/elastic/elasticsearch/pull/42662